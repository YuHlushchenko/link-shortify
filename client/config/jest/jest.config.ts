import type { Config } from 'jest'
import nextJest from 'next/jest.js'

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: './',
})

// Add any custom config to be passed to Jest
const config: Config = {
  coverageProvider: 'v8',
  testEnvironment: 'jsdom',
  clearMocks: true, // Automatically clear mock calls, instances and results before every test (looks like it doesn't)

  rootDir: '../../',
  setupFilesAfterEnv: ['<rootDir>config/jest/jest.setup.ts'], // щоб отримати доступ до ф-ї toBeInTheDocument() для тестів реакт-компонентів
  // A set of global variables that need to be available in all test environments
  //   globals: {
  //     __IS_DEV__: true,
  //   },
  // щоб працювали абсолютні імпорти в файлах з тестами
  // modulePaths: ['<rootDir>src'],

  moduleNameMapper: {
    // '^.+\\.(svg)$': 'jest-transformer-svg', // Мок для SVG як React компонента
    '^.+\\.(svg)$': '<rootDir>/src/shared/config/jest/__mocks__/fileMock.ts', // Мок для SVG як React компонента
    '\\.svg\\?url$': '<rootDir>/src/shared/config/jest/__mocks__/fileMock.ts', // Мок для SVG з параметром ?url
    'next/image': '<rootDir>/src/shared/config/jest/__mocks__/next/image.ts',
    '^@/(.*)$': '<rootDir>/src/$1', // absolute imports
    // '\\.(css|scss|sass)$': 'identity-obj-proxy', // Для CSS модулів
  },
}

// createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
export default createJestConfig(config)
