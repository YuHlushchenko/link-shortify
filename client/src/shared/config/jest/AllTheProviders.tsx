import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { NextIntlClientProvider } from 'next-intl'
import fs from 'fs'
import path from 'path'

import { IChildren, TLocale } from '@/app/types/global'

//* func for dynamic loading of all localization files for a specific language
const loadLocaleMessages = (locale: TLocale) => {
  const messages: Record<string, string> = {}
  const localePath = path.join(process.cwd(), 'public', 'locales', locale)

  // Read all files from the folder of the corresponding language
  const files = fs.readdirSync(localePath)

  files.forEach((file) => {
    if (file.endsWith('.json')) {
      const filePath = path.join(localePath, file)
      const fileContent = JSON.parse(fs.readFileSync(filePath, 'utf8'))
      Object.assign(messages, fileContent) // Join all files into one object
    }
  })

  return messages
}

const AllTheProviders = ({
  children,
  locale = 'en',
}: {
  children: IChildren
  locale?: TLocale
}) => {
  const messages = loadLocaleMessages(locale)

  return (
    <NextIntlClientProvider locale={locale} messages={messages}>
      {children}
    </NextIntlClientProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'> & { locale?: TLocale },
) => {
  const { locale = 'en', ...renderOptions } = options || {}
  return render(ui, {
    wrapper: ({ children }) => (
      <AllTheProviders locale={locale}>{children}</AllTheProviders>
    ),
    ...renderOptions,
  })
}

export * from '@testing-library/react'
export { customRender as render }
