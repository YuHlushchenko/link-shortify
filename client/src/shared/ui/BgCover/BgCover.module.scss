@use '@/app/styles/mixins.scss' as *;

%cube {
  border-radius: 60px;
  background-color: var(--bg-color);
  filter: brightness(1.1);
  transform: rotate(-45deg);

  @include for-tablet {
    border-radius: 50px;
  }
  @include for-phone {
    border-radius: 20px;
  }
}

.container {
  z-index: -1;
  position: absolute;
  min-height: 100svh;
  height: 100%;
  width: 100%;
}

.bgSwirlContainer {
  position: absolute;
  inset: 0;
  overflow: hidden;

  img {
    width: 100%;
    height: 100%;
    overflow: visible;
    object-fit: cover;
    object-position: center;
    animation: rotate-swirl 30s infinite linear;
  }
}

.bgCubesContainer {
  position: absolute;
  inset: 0;
  overflow: hidden;
}

.cubeOneContainer {
  position: absolute;
  top: 10px;
  right: 185px;
  animation: float-cube 10s infinite ease-in-out;

  @include for-small-desktop {
    right: 150px;
  }
  @include for-phone {
    top: 80px;
    right: 55px;
  }
}

.cubeOne {
  @extend %cube;

  width: 260px;
  height: 260px;
  box-shadow: 11px -20px 40px 0 var(--shadow-color);

  @include for-small-desktop {
    width: 210px;
    height: 210px;
  }
  @include for-tablet {
    width: 170px;
    height: 170px;
  }
  @include for-phone {
    width: 90px;
    height: 90px;
  }
}

.cubeTwoContainer {
  position: absolute;
  bottom: 0;
  left: 150px;
  animation: float-cube 10s 1s infinite ease-in-out;

  @include for-small-desktop {
    left: 120px;
  }
  @include for-phone {
    bottom: 190px;
    left: 0;
  }
}

.cubeTwo {
  @extend %cube;

  width: 335px;
  height: 335px;
  box-shadow: 20px -15px 45px 0 var(--shadow-color);

  @include for-small-desktop {
    width: 270px;
    height: 270px;
  }
  @include for-tablet {
    width: 220px;
    height: 220px;
  }
  @include for-phone {
    width: 130px;
    height: 130px;
  }
}

.cubeThreeContainer {
  position: absolute;
  bottom: 100px;
  right: 180px;
  animation: float-cube 10s 3s infinite ease-in-out;

  @include for-small-desktop {
    bottom: 70px;
    right: 120px;
  }
  @include for-phone {
    bottom: 90px;
    right: 25px;
  }
}

.cubeThree {
  @extend %cube;

  width: 335px;
  height: 335px;
  box-shadow: -20px 15px 45px 0 var(--shadow-color);

  @include for-small-desktop {
    width: 270px;
    height: 270px;
  }
  @include for-tablet {
    width: 220px;
    height: 220px;
  }
  @include for-phone {
    width: 110px;
    height: 110px;
  }
}

@keyframes rotate-swirl {
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
}

@keyframes float-cube {
  0% {
    transform: translateY(0) rotate(0deg);
  }

  50% {
    transform: translateY(-40px) rotate(25deg);
  }

  100% {
    transform: translateY(0) rotate(0deg);
  }
}
