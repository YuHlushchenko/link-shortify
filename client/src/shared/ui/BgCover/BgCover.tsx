import Image from 'next/image'

import styles from './BgCover.module.scss'

import SwirlUrl from 'public/assets/svgs/swirl.svg?url'

const BgCover = () => {
  return (
    <div className={styles.container}>
      <div className={styles.bgSwirlContainer}>
        <Image
          priority
          src={SwirlUrl}
          alt='swirl'
          width={400}
          height={400}
          // placeholder='blur'
          // blurDataURL={SwirlUrl}
        />
      </div>

      <div className={styles.bgCubesContainer}>
        <div className={styles.cubeOneContainer}>
          <div className={styles.cubeOne} />
        </div>
        <div className={styles.cubeTwoContainer}>
          <div className={styles.cubeTwo} />
        </div>
        <div className={styles.cubeThreeContainer}>
          <div className={styles.cubeThree} />
        </div>
      </div>
    </div>
  )
}

export default BgCover
