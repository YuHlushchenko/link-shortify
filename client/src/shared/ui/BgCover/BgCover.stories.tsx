import { Meta, StoryObj } from '@storybook/react'
import BgCover from './BgCover'
import { Theme } from '@/shared/const/theme'

const meta: Meta<typeof BgCover> = {
  title: 'Shared/BgCover',
  component: BgCover,
  argTypes: {},
}
export default meta

type Story = StoryObj<typeof BgCover>

// for dark theme
export const Default: Story = {
  args: {},
  decorators: [
    (Story) => (
      <div
        style={{
          zIndex: 1,
          width: '100%',
          minHeight: '100svh',
          height: '100%',
          position: 'relative',
          overflow: 'hidden',
        }}
      >
        <Story />
      </div>
    ),
  ],
}

// for light theme
export const DefaultLIGHT: Story = {
  args: {},
  globals: {
    theme: Theme.LIGHT,
  },
  decorators: [
    (Story) => (
      <div
        style={{
          zIndex: 1,
          width: '100%',
          minHeight: '100svh',
          height: '100%',
          position: 'relative',
          overflow: 'hidden',
        }}
      >
        <Story />
      </div>
    ),
  ],
}
