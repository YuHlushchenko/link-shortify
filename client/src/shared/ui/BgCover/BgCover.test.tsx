import React from 'react'
import { render, screen } from '@/shared/config/jest/AllTheProviders'

import BgCover from './BgCover'

describe('BgCover', () => {
  test('renders bg cover', () => {
    render(<BgCover />)

    expect(screen.getByAltText('swirl')).toBeInTheDocument()
  })

  test('matches snapshot', () => {
    const { container } = render(<BgCover />)
    expect(container).toMatchSnapshot()
  })
})
