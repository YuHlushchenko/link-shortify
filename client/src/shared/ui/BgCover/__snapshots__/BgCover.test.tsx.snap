// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BgCover matches snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="bgSwirlContainer"
    >
      <img
        alt="swirl"
        height="400"
        quality="10"
        src="test-file-stub"
        width="400"
      />
    </div>
    <div
      class="bgCubesContainer"
    >
      <div
        class="cubeOneContainer"
      >
        <div
          class="cubeOne"
        />
      </div>
      <div
        class="cubeTwoContainer"
      >
        <div
          class="cubeTwo"
        />
      </div>
      <div
        class="cubeThreeContainer"
      >
        <div
          class="cubeThree"
        />
      </div>
    </div>
  </div>
</div>
`;
